beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
+// - all indices in `indices` are available, or
+// - enough sidecars are available to trigger a reconstruction, or
+// - all peers are exhausted.
+func requestIndirectSidecarsFromPeers(
 	p DataColumnSidecarsParams,
 	roBlocks []blocks.ROBlock,
 	slotsWithCommitments map[primitives.Slot]bool,
-	missingIndicesByRoot map[[fieldparams.RootLength]byte]map[uint64]bool,
+	storedIndicesByRoot map[[fieldparams.RootLength]byte]map[uint64]bool,
+	alreadyAvailableByRoot map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn,
+	requestedIndices map[uint64]bool,
+	roots map[[fieldparams.RootLength]byte]bool,
 ) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, error) {
+	start := time.Now()
+
+	numberOfColumns := params.BeaconConfig().NumberOfColumns
+	minimumColumnCountToReconstruct := peerdas.MinimumColumnCountToReconstruct()
+
 	// Create a new random source for peer selection.
 	randomSource := rand.NewGenerator()
 
 	// Compute slots by block root.
 	slotByRoot := computeSlotByBlockRoot(roBlocks)
 
+	// For each root compute all possible data column sidecar indices excluding
+	// those already stored or already available.
+	indicesToRetrieveByRoot := make(map[[fieldparams.RootLength]byte]map[uint64]bool)
+	for root := range roots {
+		alreadyAvailableIndices := make(map[uint64]bool, len(alreadyAvailableByRoot[root]))
+		for _, sidecar := range alreadyAvailableByRoot[root] {
+			alreadyAvailableIndices[sidecar.Index] = true
+		}
+
+		storedIndices := storedIndicesByRoot[root]
+		indicesToRetrieve := make(map[uint64]bool, numberOfColumns)
+		for index := range numberOfColumns {
+			if !(storedIndices[index] || alreadyAvailableIndices[index]) {
+				indicesToRetrieve[index] = true
+			}
+		}
+
+		if len(indicesToRetrieve) > 0 {
+			indicesToRetrieveByRoot[root] = indicesToRetrieve
+		}
+	}
+
+	initialToRetrieveRootCount := len(indicesToRetrieveByRoot)
+
 	// Determine all sidecars each peers are expected to custody.
 	connectedPeersSlice := p.P2P.Peers().Connected()
 	connectedPeers := make(map[goPeer.ID]bool, len(connectedPeersSlice))
 	for _, peer := range connectedPeersSlice {
 		connectedPeers[peer] = true
 	}
 
-	indicesByRootByPeer, err := computeIndicesByRootByPeer(p.P2P, slotByRoot, missingIndicesByRoot, connectedPeers)
+	// Compute which peers have which of the missing indices.
+	indicesByRootByPeer, err := computeIndicesByRootByPeer(p.P2P, slotByRoot, indicesToRetrieveByRoot, connectedPeers)
 	if err != nil {
 		return nil, errors.Wrap(err, "explore peers")
 	}
 
-	verifiedColumnsByRoot := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
-	for len(missingIndicesByRoot) > 0 && len(indicesByRootByPeer) > 0 {
+	// Already add into results all sidecars present in `alreadyAvailableByRoot`.
+	result := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
+	for root := range roots {
+		alreadyAvailable := alreadyAvailableByRoot[root]
+		result[root] = append(result[root], alreadyAvailable...)
```

Comment:
can there be too many columns here?  do we need to check result total as we append because you append more on 451?